[InstCombine] Lower flag check pattern to use a bitmask-shift

bababuck opened the pull request:
The following C-code:
```
bool or_icmp(int type) {
  if (type == 0 || type == 6 || type == 15)
    return true;
  return false;
}
```

Currently lowers to:
```
define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp eq i32 %type, 6
  %cmp1 = icmp eq i32 %type, 0
  %or.cond = or i1 %cmp, %cmp1
  %cmp3 = icmp eq i32 %type, 15
  %or.cond1 = or i1 %cmp3, %or.cond
  ret i1 %or.cond1
}
```
But more optimally lowers to:
```
define i1 @or_icmp(i32 signext %type) {
      %0 = trunc i32 %type to i16
      %1 = lshr i16 -32703, %0
      %2 = icmp ult i32 %type, 16
      %3 = trunc i16 %1 to i1
      %or.cond1 = select i1 %2, i1 %3, i1 false
      ret i1 %or.cond1
```

Alive Proofs for the test cases:
or_icmp_7: https://alive2.llvm.org/ce/z/srzpuF

or_icmp_expand: https://alive2.llvm.org/ce/z/V-iP4Y

or_icmp_expand_trunc_type_shr: https://alive2.llvm.org/ce/z/y7U2t7

or_icmp_expand_zext_cmp: https://alive2.llvm.org/ce/z/jTBP-_

or_icmp_expand_128: https://alive2.llvm.org/ce/z/y7U2t7

or_icmp_ltu: https://alive2.llvm.org/ce/z/tcNHDo

or_icmp_3: https://alive2.llvm.org/ce/z/oA-aNe

or_icmp_i64: https://alive2.llvm.org/ce/z/Lnxq4C

or_icmp_i128: https://alive2.llvm.org/ce/z/2cf6fH

and_icmp_all_neq: https://alive2.llvm.org/ce/z/aECPEX

and_icmp_ugt: https://alive2.llvm.org/ce/z/TgNLVL

and_icmp_expand: https://alive2.llvm.org/ce/z/B6DNME

or_icmp_3_no_shift: https://alive2.llvm.org/ce/z/9Sh_uW

SimplifyCFG test cases:
bitmap_expand_deep: https://alive2.llvm.org/ce/z/WHVBnU

bababuck committed:
[InstCombine] Lower flag check pattern to use a bitmask-shift

The following C-code:

bool or_icmp(int type) {
  return type == 0 || type == 6 || type == 15;
}

Currently lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp eq i32 %type, 6
  %cmp1 = icmp eq i32 %type, 0
  %or.cond = or i1 %cmp, %cmp1
  %cmp3 = icmp eq i32 %type, 15
  %or.cond1 = or i1 %cmp3, %or.cond
  ret i1 %or.cond1
}

But more optimally lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %srl = lshr i32 32833, %type
  %srl.1 = trunc i32 %srl to i1
  %cmp = icmp ult i32 %type, 64
  %and = select i1 %cmd, i1 %srl.1, i1 false
  ret i1 %and
}

bababuck committed:
[InstCombine] Expand flag check patterns

The following C-code:

bool or_icmp(int type) {
  return type == 0 || type == 6 || type == 15;
}

Currently lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp eq i32 %type, 6
  %cmp1 = icmp eq i32 %type, 0
  %or.cond = or i1 %cmp, %cmp1
  %cmp3 = icmp eq i32 %type, 15
  %or.cond1 = or i1 %cmp3, %or.cond
  ret i1 %or.cond1
}

But more optimally lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %srl = lshr i32 32833, %type
  %srl.1 = trunc i32 %srl to i1
  %cmp = icmp ult i32 %type, 64
  %and = select i1 %cmd, i1 %srl.1, i1 false
  ret i1 %and
}

bababuck committed:
[RISCV] Lower flag check pattern to use a bitmask-shift

The following C-code:

bool or_icmp(int type) {
  return type == 0 || type == 6 || type == 15;
}

Currently lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp eq i32 %type, 6
  %cmp1 = icmp eq i32 %type, 0
  %or.cond = or i1 %cmp, %cmp1
  %cmp3 = icmp eq i32 %type, 15
  %or.cond1 = or i1 %cmp3, %or.cond
  ret i1 %or.cond1
}

But more optimally lowers to:

define i1 @or_icmp(i32 signext %type) {
entry:
  %srl = lshr i32 32833, %type
  %srl.1 = trunc i32 %srl to i1
  %cmp = icmp ult i32 %type, 64
  %and = and i1 %srl.1, %cmp
  ret i1 %and
}

bababuck committed:
[RISCV] Add new test for RISCV flag check optimization

bababuck commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -16736,6 +16736,52 @@ static SDValue combineOrAndToBitfieldInsert(SDNode *N, SelectionDAG &DAG,
 
+// or (icmp eq x, imm0), (icmp eq x, imm1) -> czero.eqz (sltui x, 64), (bext x,
+// 1 << imm0 | 1 << imm1) If [imm0, imm1] < 64
+static SDValue combineOrOfImmCmpToBitExtract(SDNode *N, SelectionDAG &DAG,
```
I would prefer to make this optimization in `InstCombine`, but I can't because:
- I don't think we have a notion of the datapath width (xlen) at the IR level, and this optimization only makes sense if the comparison immediate is less than xlen
- `lshr` poison semantics, see https://alive2.llvm.org/ce/z/9MndvW

bababuck replied:
Can use `DataLayout` to check register width.

bababuck committed:
[InstCombine] Add new tests for flag check optimization

bababuck committed:
[InstCombine] Add new tests for flag check optimization

lenary commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -16736,6 +16736,52 @@ static SDValue combineOrAndToBitfieldInsert(SDNode *N, SelectionDAG &DAG,
+
+  SmallVector<APInt, 4> FlagVals;
+  SDValue X;
+  if (!CollectSetEqImmTree(CollectSetEqImmTree, FlagVals, N, X))
```
Instead of a recursive lambda, it's more obvious just to have a static recursive function.

Are you doing anything to limit the recursion?

bababuck committed:
[RISCV] Revert implementation in RISCV

Will move to InstCombine

bababuck commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -3119,6 +3119,127 @@ static Instruction *matchFunnelShift(Instruction &Or, InstCombinerImpl &IC) {
+    if (!validImm(LHSAP) || !validImm(RHSAP))
+      return nullptr;
+    LHSAP = LHSAP.zextOrTrunc(XLen);
+    RHSAP = RHSAP.zextOrTrunc(XLen);
```
I don't know if this is the best way to wrangle various APInt's that may have differing BitWidth's- effectively I ensure they can be represented within XLen and then force them all to have a BitWidth or XLen.

bababuck committed:
Remove undef in comment to pass formatter

lukel97 commented:
Does this transform already exist somewhere in LLVM to some degree? Just checking in godbolt Clang trunk already seems to do a very similar transform on the C snippet in the PR description: https://compiler-explorer.com/z/qTbzc9e33

dtcxzyw commented:
> Does this transform already exist somewhere in LLVM to some degree? Just checking in godbolt Clang trunk already seems to do a very similar transform on the C snippet in the PR description: https://compiler-explorer.com/z/qTbzc9e33

See SimplifyCFGOpt::simplifyBranchOnICmpChain and simplifySwitchLookup

bababuck commented:
Updated the PR description.
```
bool or_icmp(int type) {
  return (type == 0 || type == 6 || type == 15);
}
```
is converted to an IR switch statement during SimplifyCFG, which SimplifyCFG then transforms into a bitmask-shift, while
```
bool or_icmp(int type) {
  if (type == 0 || type == 6 || type == 15)
    return true;
  return false;
}
```
is not optimally lowered and is addressed by this MR (see https://compiler-explorer.com/z/9P7P7bj5r). From browsing, I don't think fixing this in `SimplifyCFG` is the best solution since this pattern can exist inside a single basic block (i.e. :
```
  %cmp = icmp eq i32 %type, 0
  %cmp1 = icmp eq i32 %type, 6
  %or.cond = or i1 %cmp, %cmp1
  %cmp3 = icmp eq i32 %type, 15
  %or.cond4 = or i1 %or.cond, %cmp3
  ret i1 %or.cond4
```

dtcxzyw commented:
> From browsing, I don't think fixing this in SimplifyCFG is the best solution since this pattern can exist inside a single basic block (i.e. :

Yes. I mean we should move the common logic (ConstantComparesGatherer/switch-to-arithmetic transform) into ValueTracking and reuse it in InstCombine.

sharkautarch commented:
@bababuck 
> ...
> ```
> bool or_icmp(int type) {
>   return (type == 0 || type == 6 || type == 15);
> }
> ```
> 
> is converted to an IR switch statement during SimplifyCFG, which SimplifyCFG then transforms into a bitmask-shift, while
> 
> ```
> bool or_icmp(int type) {
>   if (type == 0 || type == 6 || type == 15)
>     return true;
>   return false;
> }
> ```
> 
> is not optimally lowered...

It looks like in the latter case, instcombine removing the phi inst prevents the function from later being optimized to a bitmask-shift during the second simplifycfg pass:
https://compiler-explorer.com/z/aca84j4Ev

bababuck committed:
[NFC] Move ConstantComparesGatherer to ValueTracking.h

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -3119,6 +3119,127 @@ static Instruction *matchFunnelShift(Instruction &Or, InstCombinerImpl &IC) {
+    if (!validImm(NewAP) || !DL.fitsInLegalInteger(BitMapAP.getActiveBits()))
+      return nullptr;
+
+    NewAP = NewAP.zextOrTrunc(XLen);
```
Well we know the value of XLen fits in an `unsigned` and we know NewAP is smaller than XLen through `validImm` so NewAP should also fit in an `unsigned` right?

bababuck committed:
[NFC] Move the logic for generating a bit-extraction sequence to ValueTracking::ConstantComparesGatherer

bababuck committed:
[NFC] Update ConstantComparesGatherer::createBitMapSeq() to take an IRBuilderBase*

Different Builder object may be passed depending on that pass

bababuck committed:
[NFC] Specify whether ConstantComparesGatherer is being used in InstCombine pass

If part of InstCombine pass, will want to require all matches have a single use.

bababuck committed:
[InstCombine] Convert series of And-ICmp.ne to bit extraction

The following:
define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp ne i32 %type, 6
  %cmp1 = icmp ne i32 %type, 0
  %and.cond = and i1 %cmp, %cmp1
  %cmp2 = icmp ne i32 %type, 15
  %and.cond1 = and i1 %cmp2, %and.cond
  ret i1 %and.cond1
}

Can more optimally lower to:
define i1 @or_icmp(i32 signext %type) {
entry:
 %trunc = trunc i32 %type to i16
 %lshr = lshr i16 -32703, %trunc
 %mask = trunc i16 %lshr to i1
 %bounds = icmp ult i32 %type, 16
 %and = select i1 %bounds, i1 %mask, i1 false
 %not = xor i1 %and, true
 ret i1 %not
}

and_icmp_all_neq: https://alive2.llvm.org/ce/z/DQdDec

bababuck committed:
[InstCombine] Allow ConstantComparesGatherer to match `ult` comparisons with no limit

If converting to switch statement want to limit the number of cases, but don't have that
concern in InstCombine since we know we are converting to a bitmap.

and_icmp_ugt: https://alive2.llvm.org/ce/z/LE7_kz

bababuck committed:
[InstCombine] Convert series of Or-ICmp.eq to bit extraction

The following:
define i1 @or_icmp(i32 signext %type) {
entry:
  %cmp = icmp eq i32 %type, 6
  %cmp1 = icmp eq i32 %type, 0
  %or.cond = or i1 %cmp, %cmp1
  %cmp2 = icmp eq i32 %type, 15
  %or.cond1 = or i1 %cmp2, %or.cond
  ret i1 %or.cond1
}

Can more optimally lower to:
define i1 @or_icmp(i32 signext %type) {
entry:
  %0 = trunc i32 %type to i16
  %1 = lshr i16 -32703, %0
  %2 = icmp ult i32 %type, 16
  %3 = trunc i16 %1 to i1
  %or.cond1 = select i1 %2, i1 %3, i1 false
  ret i1 %or.cond1
}

or_icmp_ltu: https://alive2.llvm.org/ce/z/tcNHDo

or_icmp_3: https://alive2.llvm.org/ce/z/oA-aNe

or_icmp_i64: https://alive2.llvm.org/ce/z/Lnxq4C

or_icmp_i128: https://alive2.llvm.org/ce/z/2cf6fH

bababuck committed:
[InstCombine] Update ConstantComparesGatherer::matchInstruction() to recognize existing bit-flag sequence

or_icmp_7: https://alive2.llvm.org/ce/z/srzpuF

or_icmp_expand: https://alive2.llvm.org/ce/z/V-iP4Y

or_icmp_expand_trunc_type_shr: https://alive2.llvm.org/ce/z/y7U2t7

or_icmp_expand_zext_cmp: https://alive2.llvm.org/ce/z/jTBP-_

or_icmp_expand_128: https://alive2.llvm.org/ce/z/y7U2t7

bababuck committed:
Revert prior standalone changes, will redo to reuse logic from SimplifyCFG

bababuck commented:
Moved `ConstantComparesGatherer` from `SimplifyCFG` to `ValueTracking` and re-used that pattern matching from `InstCombine`. I also pulled out the logic for `createBitMapSeq()` and moved that into `ConstantComparesGatherer` (I think this is the wrong place since this logic is for generating a new IR sequence, not Analysis so might not belong in `Analysis/ValueTracking.cpp`, so if someone knows a better spot please let me know!).

Once this is ready I can break up this MR into a few different ones for merging if desired.

bababuck committed:
[InstCombine] Move combineAndOrOfImmCmpToBitExtract() earlier in visitAnd()

Allows the BitMap to be expanded before the Not-And sequence is simplified.

and_icmp_expand: https://alive2.llvm.org/ce/z/cYgCwC

bababuck commented:
Updated proofs and added new ones for `And-ICmp.ne` tests.

bababuck committed:
Change undef to poison in comment for linter

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,59 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  unsigned MaxVal = 0;
+  // TODO: Handle case where some values are too large for map but some are not.
+  for (auto *CI : Values) {
+    unsigned Val = CI->getValue().getLimitedValue();
```
`getLimitedValue` returns a 64-bit value.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,59 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  // If the maximum value in the bitmap is larger than can be stored
+  // in Index, don't have to worry about overflow on the shift
+  unsigned IndexBits = dyn_cast<IntegerType>(Index->getType())->getBitWidth();
+  if (MaxVal >= (1u << IndexBits)) {
```
`1u << IndexBits` may overflow (In C++ it is an immediate UB). I'd suggest using `llvm::isIntN` instead.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,59 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  }
+
+  if (!ConstantCompare.IsEq)
+    Result = Builder.CreateNot(Result);
```
Can we invert the bitmap and the default value for out-of-bounds cases instead?

bababuck committed:
[ValueTracking] Pass a reference to Builder in createBitMapSeq rather than a pointer

bababuck committed:
[InstCombine] Use getScalarSizeInBits rather than getBitWidth

For handling vector types.

bababuck committed:
Remove un-needed signext from test

bababuck committed:
[InstCombine] Use uint64_t rather than unsigned for holding result of getLimitedValue()

getLimitedValue() returns a 64 bit value

bababuck committed:
[InstCombine] Invert the BitMap and select default value rather than inverting the result

Saves an instruction.

and_icmp_all_neq: https://alive2.llvm.org/ce/z/aECPEX

and_icmp_ugt: https://alive2.llvm.org/ce/z/TgNLVL

and_icmp_expand: https://alive2.llvm.org/ce/z/B6DNME

bababuck committed:
[InstCombine] Fix condition for deciding whether or not too add bounds check

Uses isUIntN()

Before, was logic bug, only working because of overflow behavior.

bababuck committed:
[InstCombine] Don't try to simplify flag checks of vector instructions

TODO: Support vectors.

bababuck committed:
[InstCombine] Add vector test to flag_check.ll

bababuck committed:
[ValueTracking] Assert that ConstantComparesGatherer doesn't operate on vector types

Will support in future PR.

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,65 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  // If the maximum value in the bitmap is larger than can be stored
+  // in Index, don't have to worry about overflow on the shift
+  unsigned IndexBits =
+      dyn_cast<IntegerType>(Index->getType())->getScalarSizeInBits();
```
If you use `dyn_cast` you must check the result for `null`. If it can't fail, use `cast`.

Also there's no need to cast if you're going to call `getScalarSizeInBits()` instead of `getBitWidth()`

bababuck replied:
Thanks!

bababuck committed:
[InstCombine] Remove un-needed cast.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,64 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+
+  // TODO: Handle ConstantCompare.Extra case
+  // If expanding an existing case, only adding one extra case is still good
+  if (!Index || !isGuaranteedNotToBeUndefOrPoison(Index) ||
```
Is `isGuaranteedNotToBeUndefOrPoison` needed? If the index is poison, the whole expression is also poison.

bababuck replied:
I think that's correct, but I think it just needs to be softened to `isGuarateedNotToBeUndefOrPoison()`, since if `Index` is `undef`, it can cause `poison` in the shift operation. See https://alive2.llvm.org/ce/z/7qXpr5

bababuck committed:
[InstCombine] Add new test for overflow case

See https://github.com/llvm/llvm-project/pull/169557#discussion_r2679517546.

bababuck committed:
[InstCombine] Fix overflow bug, add test

https://alive2.llvm.org/ce/z/9Sh_uW

bababuck commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,64 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  // in Index, don't have to worry about overflow on the shift
+  unsigned IndexBits = Index->getType()->getScalarSizeInBits();
+  if (isUIntN(IndexBits, MaxVal)) {
+    Value *MaxValue = ConstantInt::get(Context, APInt(IndexBits, MaxVal + 1));
```
Fixed, added test.

bababuck committed:
Soften isGuaranteedNotToBeUndefOrPoison() to isGuaranteedNotToBeUndef()

andjo403 commented on llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:
```diff
@@ -2410,6 +2410,64 @@ Value *InstCombinerImpl::reassociateBooleanAndOr(Value *LHS, Value *X, Value *Y,
+  // TODO: Handle ConstantCompare.Extra case
+  // If expanding an existing case, only adding one extra case is still good
+  if (!Index || !isGuaranteedNotToBeUndef(Index) ||
+      (ConstantCompare.UsedICmps + ConstantCompare.ExpansionCase) < 3 ||
```
if ExpansionCase is set do that not mean that we have already folded 3 icmps and it shall be enough with one more icmp for the fold to be profitable?

bababuck committed:
Use forward declaration of IRBuilderBase rather than including IRBuilder.h

bababuck committed:
Rename InstCombine to OneUseOnly

More closely matches the main use of the variable

bababuck committed:
Pull out MaxSpan variable

bababuck committed:
Rename Or to I

Can be And or Or.

bababuck committed:
Pre-commit SimplifyCFG tests for the expansion case

Use an `opt` binary from the root of this PR to initialize the test
output.

bababuck committed:
Update SimplifyCFG test for the changes in this MR

bitmap_expand_deep: https://alive2.llvm.org/ce/z/WHVBnU

bababuck committed:
Use isIntegerTy(1) rather than isVectorTy()

Only want to considere And/Or instructions using type i1.
